libraries/standard/mqtt/src/iot_mqtt_subscription.c

```diff
@@ -88,108 +142,123 @@ static bool _packetMatch( const IotLink_t * pSubscriptionLink,
+        status = true;
     }
 
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool _topicFilterMatch( const char * pTopicName,
+                               uint16_t topicNameLength,
+                               const char * pTopicFilter,
+                               uint16_t topicFilterLength )
+{
+    bool status = false;
+    uint16_t nameIndex = 0, filterIndex = 0;
+
     while( ( nameIndex < topicNameLength ) && ( filterIndex < topicFilterLength ) )
     {
         /* Check if the character in the topic name matches the corresponding
          * character in the topic filter string. */
         if( pTopicName[ nameIndex ] == pTopicFilter[ filterIndex ] )
         {
             /* Handle special corner cases as documented by the MQTT protocol spec. */
+            status = _matchTerminalWildcards( pTopicFilter,
+                                              nameIndex,
+                                              filterIndex,
+                                              topicNameLength,
+                                              topicFilterLength );
 
-            /* Filter "sport/#" also matches "sport" since # includes the parent level. */
-            if( nameIndex == topicNameLength - 1 )
-            {
-                if( filterIndex == topicFilterLength - 3 )
-                {
-                    if( pTopicFilter[ filterIndex + 1 ] == '/' )
-                    {
-                        if( pTopicFilter[ filterIndex + 2 ] == '#' )
-                        {
-                            status = true;
-                            goto cleanup;
-                        }
-                    }
-                }
-            }
-
-            /* Filter "sport/+" also matches the "sport/" but not "sport". */
-            if( nameIndex == topicNameLength - 1 )
+            if( status == true )
             {
-                if( filterIndex == topicFilterLength - 2 )
-                {
-                    if( pTopicFilter[ filterIndex + 1 ] == '+' )
-                    {
-                        status = true;
-                        goto cleanup;
-                    }
-                }
+                goto cleanup;
             }
         }
         else
         {
-            /* Check for wildcards. */
-            if( pTopicFilter[ filterIndex ] == '+' )
-            {
-                /* Move topic name index to the end of the current level.
-                 * This is identified by '/'. */
-                while( nameIndex < topicNameLength && pTopicName[ nameIndex ] != '/' )
-                {
-                    nameIndex++;
-                }
-
-                /* Increment filter index to skip '/'. */
-                filterIndex++;
-                continue;
-            }
-            else if( pTopicFilter[ filterIndex ] == '#' )
+            /* Check for matching wildcards. */
+            if( _matchWildcards( pTopicFilter,
```

Comment:
This is different code structure from above, where you assign the function output to a local, then test it.

Reply:
The behavior of these functions is slightly different:
- `_matchTerminalWildcards` only verifies positive matches, so it can exit on all true values.
- `_matchWildcards` may determine a positive or negative match; it needs a return value to determine whether it should exit, and the status for exit.

I'll think about how to make these more similar...